Sming/Wiring/WHashMap.h

```diff
@@ -85,11 +85,13 @@ class HashMap
     */
     const K& keyAt(unsigned int idx) const
     {
+      assert(idx < count());
```

Comment:
This whole PR is to improve the lookup but can you check in practise if that is the case?
Let's say that we usually have an instance of HttpHeaders with seven standard headers plus one or two custom headers. Can you benchmark the old implementation and this one and paste the results here?

Reply:
Test code:
```
#include "Network/Http/HttpHeaders.h"
#include "Network/WebConstants.h"
#include "Services/Profiling/ElapseTimer.h"

static void printHeaders(const HttpHeaders& headers)
{
	debugf("  count: %d", headers.count());
	for(unsigned i = 0; i < headers.count(); ++i) {
		String s = headers[i];
		debug_hex(INFO, "  ", s.c_str(), s.length(), 0, 32);
	}
}

void testHttpHeaders()
{
	HttpHeaders headers;
	headers["X-Auth-Header"] = "Value Data";
	headers["x-Auth-header"] = "Value Data New";

	debugf("Header: %s", headers["X-AUTH-HEADER"].c_str());
	printHeaders(headers);

	headers.remove("x-auth-header");
	printHeaders(headers);

	headers[HTTP_HEADER_ACCESS_CONTROL_ALLOW_ORIGIN] = "This is access control origin";

	debugf("Header: %s", headers["access-control-Allow-Origin"].c_str());

	debugf("Headers count: %d", headers.count());

	headers.clear();
	printHeaders(headers);

	const HttpHeaders& constHeaders = headers;
	debugf("Non-Existent Header (const): \"%s\"", constHeaders["Non-Existent"].c_str());
	printHeaders(headers);

	debugf("Non-Existent Header (non-const): \"%s\"", headers["Non-Existent"].c_str());
	printHeaders(headers);

	headers["X-Auth-Header"] = "Value Data";
	debugf("set(\"X-Auth-Header\")");
	printHeaders(headers);

	headers["x-Auth-header"] = "Value Data New";
	debugf("set(\"x-Auth-Header\")");
	printHeaders(headers);

	debugf("Header[\"X-AUTH-HEADER\"]: \"%s\"", headers["X-AUTH-HEADER"].c_str());
	printHeaders(headers);

	headers.remove("x-auth-header");
	debugf("headers.remove(\"x-auth-header\")");
	printHeaders(headers);

	headers["Mary"] = "Had a little lamb";
	headers[HTTP_HEADER_CONTENT_LENGTH] = "12345";
	headers["George"] = "Was an idiot";
	printHeaders(headers);

	debugf("\nPROFILING");

	// Set header values
	auto freeHeap = system_get_free_heap_size();
	headers.clear();
	ElapseTimer timer;
	headers[HTTP_HEADER_CONTENT_ENCODING] = _F("gzip");
	headers[HTTP_HEADER_CONTENT_LENGTH] = 6042;
	headers[HTTP_HEADER_ETAG] = "00f-3d-179a0-0";
	headers[HTTP_HEADER_CONNECTION] = _F("keep-alive");
	headers[HTTP_HEADER_SERVER] = _F("HttpServer/Sming");
	headers[HTTP_HEADER_CONTENT_TYPE] = ContentType::toString(MIME_JS);
	headers[HTTP_HEADER_CACHE_CONTROL] = "max-age=31536000, public";
	headers[HTTP_HEADER_ACCESS_CONTROL_ALLOW_ORIGIN] = "*";
	auto standardElapsed = timer.elapsed();
	headers[F("X-Served-By")] = _F("Donkey Kong");
	headers[F("Vary")] = _F("Accept-Encoding");
	headers[F("X-Fastly-Request-ID")] = _F("38ef411e0ec3bf681d29d8b4b51f3516d3ef9e03");
	auto totalElapsed = timer.elapsed();
	debugf("Set header values");
	debugf("  Elapsed standard: %u, total: %u, heap used: %u", standardElapsed, totalElapsed,
		   freeHeap - system_get_free_heap_size());

	// Query header value by field name
	auto queryByEnum = [&](HttpHeaderFieldName name) {
		debugf("  header[\"%s\"]: %s", headers.toString(name).c_str(), headers[name].c_str());
	};
	auto queryByString = [&](const String& name) {
		debugf("  header[\"%s\"]: %s", name.c_str(), headers[name].c_str());
	};
	debugf("Query header values");
	timer.start();
	queryByEnum(HTTP_HEADER_CONTENT_ENCODING);
	queryByEnum(HTTP_HEADER_CONTENT_LENGTH);
	queryByEnum(HTTP_HEADER_ETAG);
	queryByEnum(HTTP_HEADER_CONNECTION);
	queryByEnum(HTTP_HEADER_SERVER);
	queryByEnum(HTTP_HEADER_CONTENT_TYPE);
	queryByEnum(HTTP_HEADER_CACHE_CONTROL);
	queryByEnum(HTTP_HEADER_ACCESS_CONTROL_ALLOW_ORIGIN);
	standardElapsed = timer.elapsed();
	queryByString("X-Served-By");
	queryByString("Vary");
	queryByString("X-Fastly-Request-ID");
	totalElapsed = timer.elapsed();
	debugf("  Elapsed standard: %u, total: %u", standardElapsed, totalElapsed);

	// Print header values - accessed by index
	debugf("Printing %u headers", headers.count());
	timer.start();
	printHeaders(headers);
	debugf("  Elapsed: %u", timer.elapsed());
}
```

Reply:
# Comparing old and new versions

Tested with 8 standard fields and 3 custom. Breakdown of results:
* Store standard value: -37us
* Store custom value: +40us
* Query standard value: -29us
* Query custom value: +12us
* Reading out headers: +18us per line
* Heap used: -224 bytes (30% reduction)

I took a longer look at the times to read out lines, and realised a further optimisation could be made when concatenating the key/value, so applied that to both versions. It shaves about 900us off the print time (82us each). Note that the serial printing uses an 8K buffer to minimise impact on timings. I repeated each test and timings vary only slightly.


## HashMap(String, String)

```
PROFILING
101965 Set header values
102028   Elapsed standard: 742, total: 1105, heap used: 760
102139 Query header values
102198   header["Content-Encoding"]: gzip
102325   header["Content-Length"]: 6042
102443   header["ETag"]: 00f-3d-179a0-0
102560   header["Connection"]: keep-alive
102682   header["Server"]: HttpServer/Sming
102804   header["Content-Type"]: text/javascript
102931   header["Cache-Control"]: max-age=31536000, public
103068   header["Access-Control-Allow-Origin"]: *
103224   header["X-Served-By"]: Donkey Kong
103326   header["Vary"]: Accept-Encoding
103429   header["X-Fastly-Request-ID"]: 38ef411e0ec3bf681d29d8b4b51f3516d3ef9e03
103567   Elapsed standard: 1009, total: 1355
103658 Printing 11 headers
103716   count: 11
  : 00000000 43 6f 6e 74 65 6e 74 2d 45 6e 63 6f 64 69 6e 67 3a 20 67 7a 69 70 0d 0a                          Content-Encoding: gzip..
  : 00000000 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 3a 20 36 30 34 32 0d 0a                                Content-Length: 6042..
  : 00000000 45 54 61 67 3a 20 30 30 66 2d 33 64 2d 31 37 39 61 30 2d 30 0d 0a                                ETag: 00f-3d-179a0-0..
  : 00000000 43 6f 6e 6e 65 63 74 69 6f 6e 3a 20 6b 65 65 70 2d 61 6c 69 76 65 0d 0a                          Connection: keep-alive..
  : 00000000 53 65 72 76 65 72 3a 20 48 74 74 70 53 65 72 76 65 72 2f 53 6d 69 6e 67 0d 0a                    Server: HttpServer/Sming..
  : 00000000 43 6f 6e 74 65 6e 74 2d 54 79 70 65 3a 20 74 65 78 74 2f 6a 61 76 61 73 63 72 69 70 74 0d 0a     Content-Type: text/javascript..
  : 00000000 43 61 63 68 65 2d 43 6f 6e 74 72 6f 6c 3a 20 6d 61 78 2d 61 67 65 3d 33 31 35 33 36 30 30 30 2c  Cache-Control: max-age=31536000,
    00000020 20 70 75 62 6c 69 63 0d 0a                                                                        public..
  : 00000000 41 63 63 65 73 73 2d 43 6f 6e 74 72 6f 6c 2d 41 6c 6c 6f 77 2d 4f 72 69 67 69 6e 3a 20 2a 0d 0a  Access-Control-Allow-Origin: *..
  : 00000000 58 2d 53 65 72 76 65 64 2d 42 79 3a 20 44 6f 6e 6b 65 79 20 4b 6f 6e 67 0d 0a                    X-Served-By: Donkey Kong..
  : 00000000 56 61 72 79 3a 20 41 63 63 65 70 74 2d 45 6e 63 6f 64 69 6e 67 0d 0a                             Vary: Accept-Encoding..
  : 00000000 58 2d 46 61 73 74 6c 79 2d 52 65 71 75 65 73 74 2d 49 44 3a 20 33 38 65 66 34 31 31 65 30 65 63  X-Fastly-Request-ID: 38ef411e0ec
    00000020 33 62 66 36 38 31 64 32 39 64 38 62 34 62 35 31 66 33 35 31 36 64 33 65 66 39 65 30 33 0d 0a     3bf681d29d8b4b51f3516d3ef9e03..
108078   Elapsed: 4374
```


## HashMap(enum, String)

```
PROFILING
103698 Set header values
103761   Elapsed standard: 483, total: 964, heap used: 536
103869 Query header values
103933   header["Content-Encoding"]: gzip
104028   header["Content-Length"]: 6042
104119   header["ETag"]: 00f-3d-179a0-0
104211   header["Connection"]: keep-alive
104304   header["Server"]: HttpServer/Sming
104397   header["Content-Type"]: text/javascript
104497   header["Cache-Control"]: max-age=31536000, public
104604   header["Access-Control-Allow-Origin"]: *
104726   header["X-Served-By"]: Donkey Kong
104848   header["Vary"]: Accept-Encoding
104974   header["X-Fastly-Request-ID"]: 38ef411e0ec3bf681d29d8b4b51f3516d3ef9e03
105136   Elapsed standard: 780, total: 1195
105227 Printing 11 headers
105283   count: 11
  : 00000000 43 6f 6e 74 65 6e 74 2d 45 6e 63 6f 64 69 6e 67 3a 20 67 7a 69 70 0d 0a                          Content-Encoding: gzip..
  : 00000000 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 3a 20 36 30 34 32 0d 0a                                Content-Length: 6042..
  : 00000000 45 54 61 67 3a 20 30 30 66 2d 33 64 2d 31 37 39 61 30 2d 30 0d 0a                                ETag: 00f-3d-179a0-0..
  : 00000000 43 6f 6e 6e 65 63 74 69 6f 6e 3a 20 6b 65 65 70 2d 61 6c 69 76 65 0d 0a                          Connection: keep-alive..
  : 00000000 53 65 72 76 65 72 3a 20 48 74 74 70 53 65 72 76 65 72 2f 53 6d 69 6e 67 0d 0a                    Server: HttpServer/Sming..
  : 00000000 43 6f 6e 74 65 6e 74 2d 54 79 70 65 3a 20 74 65 78 74 2f 6a 61 76 61 73 63 72 69 70 74 0d 0a     Content-Type: text/javascript..
  : 00000000 43 61 63 68 65 2d 43 6f 6e 74 72 6f 6c 3a 20 6d 61 78 2d 61 67 65 3d 33 31 35 33 36 30 30 30 2c  Cache-Control: max-age=31536000,
    00000020 20 70 75 62 6c 69 63 0d 0a                                                                        public..
  : 00000000 41 63 63 65 73 73 2d 43 6f 6e 74 72 6f 6c 2d 41 6c 6c 6f 77 2d 4f 72 69 67 69 6e 3a 20 2a 0d 0a  Access-Control-Allow-Origin: *..
  : 00000000 58 2d 53 65 72 76 65 64 2d 42 79 3a 20 44 6f 6e 6b 65 79 20 4b 6f 6e 67 0d 0a                    X-Served-By: Donkey Kong..
  : 00000000 56 61 72 79 3a 20 41 63 63 65 70 74 2d 45 6e 63 6f 64 69 6e 67 0d 0a                             Vary: Accept-Encoding..
  : 00000000 58 2d 46 61 73 74 6c 79 2d 52 65 71 75 65 73 74 2d 49 44 3a 20 33 38 65 66 34 31 31 65 30 65 63  X-Fastly-Request-ID: 38ef411e0ec
    00000020 33 62 66 36 38 31 64 32 39 64 38 62 34 62 35 31 66 33 35 31 36 64 33 65 66 39 65 30 33 0d 0a     3bf681d29d8b4b51f3516d3ef9e03..
109825   Elapsed: 4555
```

Reply:
Just tried it with Vector<String> instead of CStringArray. Interesting. Note the heap usage is rather less, implies that CStringArray is inducing fragmentation. One of the issues with using the String is that it doesn't inherently reserve any more space than necessary for the operation at hand. For example, `String s = "a" + "b" + "c"` will require (at least) three reallocations. We could anticipate this kind of typical usage by allocating in multiples of, say, 32 bytes to allow a little room. I'm having a play with CStringArray to see if I can strike a balance for use with HttpHeaders, otherwise `Vector<String>` might be a better option...

## Using Vector<String> instead of CStringArray

```
PROFILING
107841 Set header values
107907   Elapsed standard: 493, total: 935, heap used: 232
108012 Query header values
108076   header["Content-Encoding"]: gzip
108170   header["Content-Length"]: 6042
108263   header["ETag"]: 00f-3d-179a0-0
108355   header["Connection"]: keep-alive
108448   header["Server"]: HttpServer/Sming
108545   header["Content-Type"]: text/javascript
108643   header["Cache-Control"]: max-age=31536000, public
108752   header["Access-Control-Allow-Origin"]: *
108874   header["X-Served-By"]: Donkey Kong
109000   header["Vary"]: Accept-Encoding
109128   header["X-Fastly-Request-ID"]: 38ef411e0ec3bf681d29d8b4b51f3516d3ef9e03
109285   Elapsed standard: 783, total: 1197
109374 Printing 11 headers
109432   count: 11
  : 00000000 43 6f 6e 74 65 6e 74 2d 45 6e 63 6f 64 69 6e 67 3a 20 67 7a 69 70 0d 0a                          Content-Encoding: gzip..
  : 00000000 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 3a 20 36 30 34 32 0d 0a                                Content-Length: 6042..
  : 00000000 45 54 61 67 3a 20 30 30 66 2d 33 64 2d 31 37 39 61 30 2d 30 0d 0a                                ETag: 00f-3d-179a0-0..
  : 00000000 43 6f 6e 6e 65 63 74 69 6f 6e 3a 20 6b 65 65 70 2d 61 6c 69 76 65 0d 0a                          Connection: keep-alive..
  : 00000000 53 65 72 76 65 72 3a 20 48 74 74 70 53 65 72 76 65 72 2f 53 6d 69 6e 67 0d 0a                    Server: HttpServer/Sming..
  : 00000000 43 6f 6e 74 65 6e 74 2d 54 79 70 65 3a 20 74 65 78 74 2f 6a 61 76 61 73 63 72 69 70 74 0d 0a     Content-Type: text/javascript..
  : 00000000 43 61 63 68 65 2d 43 6f 6e 74 72 6f 6c 3a 20 6d 61 78 2d 61 67 65 3d 33 31 35 33 36 30 30 30 2c  Cache-Control: max-age=31536000,
    00000020 20 70 75 62 6c 69 63 0d 0a                                                                        public..
  : 00000000 41 63 63 65 73 73 2d 43 6f 6e 74 72 6f 6c 2d 41 6c 6c 6f 77 2d 4f 72 69 67 69 6e 3a 20 2a 0d 0a  Access-Control-Allow-Origin: *..
  : 00000000 58 2d 53 65 72 76 65 64 2d 42 79 3a 20 44 6f 6e 6b 65 79 20 4b 6f 6e 67 0d 0a                    X-Served-By: Donkey Kong..
  : 00000000 56 61 72 79 3a 20 41 63 63 65 70 74 2d 45 6e 63 6f 64 69 6e 67 0d 0a                             Vary: Accept-Encoding..
  : 00000000 58 2d 46 61 73 74 6c 79 2d 52 65 71 75 65 73 74 2d 49 44 3a 20 33 38 65 66 34 31 31 65 30 65 63  X-Fastly-Request-ID: 38ef411e0ec
    00000020 33 62 66 36 38 31 64 32 39 64 38 62 34 62 35 31 66 33 35 31 36 64 33 65 66 39 65 30 33 0d 0a     3bf681d29d8b4b51f3516d3ef9e03..
114079   Elapsed: 4660

```

Reply:
The results above indicate that going with HashMap(enum, String) reduces the overall time and consumes less heap, for the example given. In actual use I think we'll get more improvement because the Write/Query operations are typically used multiple times, whereas the Read operation is performed only once.
I'm happy with the code now!